packages/@uppy/aws-s3-multipart/src/MultipartUploader.js

```diff
@@ -162,39 +162,48 @@ class MultipartUploader {
       return
     }
 
-    // For a 100MB file, with the default min chunk size of 5MB and a limit of 10:
-    //
-    // Total 20 parts
-    // ---------
-    // Need 1 is 10
-    // Need 2 is 5
-    // Need 3 is 5
-    const need = this.options.limit - this.partsInProgress
-    const completeChunks = this.chunkState.filter((state) => state.done).length
-    const remainingChunks = this.chunks.length - completeChunks
-    let minNeeded = Math.ceil(this.options.limit / 2)
-    if (minNeeded > remainingChunks) {
-      minNeeded = remainingChunks
-    }
-    if (need < minNeeded) return
+    const getChunkIndexes = () => {
+      // For a 100MB file, with the default min chunk size of 5MB and a limit of 10:
+      //
+      // Total 20 parts
+      // ---------
+      // Need 1 is 10
+      // Need 2 is 5
+      // Need 3 is 5
+      const need = this.options.limit - this.partsInProgress
+      const completeChunks = this.chunkState.filter((state) => state.done).length
+      const remainingChunks = this.chunks.length - completeChunks
+      let minNeeded = Math.ceil(this.options.limit / 2)
+      if (minNeeded > remainingChunks) {
+        minNeeded = remainingChunks
+      }
+      if (need < minNeeded) return []
 
-    const candidates = []
-    for (let i = 0; i < this.chunkState.length; i++) {
-      const state = this.chunkState[i]
-      // eslint-disable-next-line no-continue
-      if (state.done || state.busy) continue
+      const chunkIndexes = []
+      for (let i = 0; i < this.chunkState.length; i++) {
+        const state = this.chunkState[i]
+        // eslint-disable-next-line no-continue
+        if (state.done || state.busy) continue
 
-      candidates.push(i)
-      if (candidates.length >= need) {
-        break
+        chunkIndexes.push(i)
+        if (chunkIndexes.length >= need) {
+          break
+        }
       }
+
+      return chunkIndexes
     }
-    if (candidates.length === 0) return
 
-    this.#prepareUploadParts(candidates).then((result) => {
-      candidates.forEach((index) => {
+    const chunkIndexes = getChunkIndexes()
+
+    if (chunkIndexes.length === 0) return
+
+    this.#prepareUploadParts(chunkIndexes).then((result) => {
```

Comment:
I think we should return here, otherwise the potential promise rejection can never be handled.
```suggestion
    return this.#prepareUploadParts(chunkIndexes).then((result) => {
```

Reply:
I can do it, for it's actually not required. `prepareUploadParts` uses `retryable` which throws an error at some point. The error is catched in `createUpload`, where this function is called.

Reply:
even better: use async/await for more readable code?

Reply:
Can't do that because it uses `retryable` under the hood which gives the error as a second callback. Similar to the one below:

```js
        this.#uploadPartRetryable(index, prePreparedPart).then(() => {
          this.#uploadParts()
        }, (err) => {
        }
```

Reply:
If it doesn't hurt, let's return the promise, it might be of great help in a future debugging session in the future.

Reply:
I meant like this:
```js
const { presignedUrls, headers } = await this.#prepareUploadParts(chunkIndexes)

chunkIndexes.forEach((index) => {
  const partNumber = index + 1
  const prePreparedPart = { url: presignedUrls[partNumber], headers: headers?.[partNumber] }
  this.#uploadPartRetryable(index, prePreparedPart).then(() => {
    this.#uploadParts()
  }, (err) => {
    this.#onError(err)
  })
})
```